22.0.0.11-beta draft (#2571)
OpenLiberty/blogs: posts/2022-09-29-22.0.0.11-beta.adoc

---
layout: post
title: "InstantOn With Checkpoint and Restore in Open liberty 22.0.0.11-beta!"
# Do NOT change the categories section
categories: blog
author_picture: https://avatars3.githubusercontent.com/reecenana
author_github: https://github.com/reecenana
seo-title: InstantOn With Checkpoint and Restore in Open liberty 22.0.0.11-beta! - OpenLiberty.io
seo-description: InstantOn will have your applications starting at light speed! Developer experience and performance enhancements come to Jakarta EE 10 for mail, persistence and connectors.
blog_description: InstantOn will have your applications starting at light speed! Developer experience and performance enhancements come to Jakarta EE 10 for mail, persistence and connectors.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= InstantOn With Checkpoint and Restore in Open liberty 22.0.0.11-beta!
Reece Nana <https://github.com/reecenana>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

// // // // // // // //

> Open Liberty 22.0.0.11-beta provides an exciting new capability called InstantOn which allows applications to checkpoint the startup process and restore the application process very quickly, providing instant on behavior for your application.

Reviewer: Is checkpoint and restore a familiar concept for most users? If not, we could focus on the value add for the intro.

```suggestion
Open Liberty 22.0.0.11-beta provides an exciting new capability called InstantOn, which provides incredibly fast startup times and can reduce the overall cloud costs for deployed applications.
```

> Jakarta EE 10 has improved support provided across features such as mail, persistence and connectors to enhance users performance

Reviewer: ```suggestion
Jakarta EE 10 has improved support provided across features such as mail, persistence, and connectors to enhance user performance
```

> and experience when developing.

Reviewer: ```suggestion
and developer experience.
```

> With the new Jakarta Mail 2.1 feature, you can now entirely remove direct dependency of the specification API's on the reference implementations, reducing the time taken to compile your Jakarta code. Open Liberty is keeping our connectors implementation up to date with the latest specification.

Reviewer: ```suggestion
With the new Jakarta Mail 2.1 feature, you can now entirely remove  the specification API's direct dependency on the reference implementations, reducing the time it takes to compile your Jakarta code. Open Liberty is keeping our connectors implementation up to date with the latest specification.
```

> Also, the new Open Liberty servlet 6.0 feature provides a new ability to set any attribute using a cookie's setAttribute method! 

Reviewer: ```suggestion
Also, the new Open Liberty servlet-6.0 feature provides the ability to set any attribute using a cookie's `setAttribute` method.
```

// // // // // // // //

The link:{url-about}[Open Liberty] 22.0.0.11-beta includes the following beta features (along with link:{url-prefix}/docs/latest/reference/feature/feature-overview.html[all GA features]):

* <<instanton, InstantOn With Checkpoint and Restore>>
* <<mail, Jakarta Mail 2.1>>
* <<support,  Jakarta EE 10: Support for Jakarta Servlet 6.0>>
* <<connectors, Connectors-2.1>>
* <<persistence, Persistence 3.1>>

// // // // // // // //

// https://github.com/OpenLiberty/open-liberty/issues/21977
[#instanton]
== InstantOn With Checkpoint and Restore

> Traditionally, starting an application can take several seconds and sometimes minutes before being able to service the application's first request. Longer startup times inhibit an application's ability to scale to zero when deployed in a cloud environment such as Kubernetes, which in turn can drive up costs. 

Reviewer: ```suggestion
Traditionally, starting an application can take several seconds, and sometimes minutes, before the application can service its first request. Longer startup times inhibit an application's ability to scale to zero when deployed in a cloud environment such as Kubernetes, which in turn can drive up costs. 
```

> Liberty's InstantOn capability is an innovative way to drastically reduce that startup through Checkpoint and Restore. Restoring an application from a checkpoint process can take much less time, up to 10 times faster than starting the Liberty application from the beginning.

Reviewer: ```suggestion
Liberty's InstantOn capability is an innovative way to drastically reduce that startup time by usinglink:https://criu.org/Main_Page[Checkpoint and Restore]. Restoring an application from a checkpoint process can be up to 10 times faster than starting the Liberty application from the beginning.
```

For more information, see the link:{url-prefix}/blog/2022/09/27/instant-on-220011-beta.html[Liberty InstantOn startup for cloud native Java applications] blog post.

// https://github.com/OpenLiberty/open-liberty/issues/22423
[#mail]
== Jakarta Mail 2.1

Jakarta Mail 2.1 support allows your Jakarta EE 10 applications to send, read, and edit E-Mail from mail servers via the Jakarta Mail APIs.

> The biggest change to Jakarta Mail 2.1 specification is the removal of the direct dependency of the specification APIs on the reference implementation. This means that in certain cases there will no longer be a need to compile your Jakarta Mail code against both the specification and the reference implementation (provided you aren't making any direct calls to the RI in your code).

Reviewer: ```suggestion
The biggest change to the Jakarta Mail 2.1 specification is the removal of the direct dependency of the specification APIs on the reference implementation. This means that, in certain cases, there is no longer a need to compile your Jakarta Mail code against both the specification and the reference implementation (provided you aren't making any direct calls to the RI in your code).
```

>  In order to facilitate this, Jakarta Mail 2.1 also introduced a utility interface link:https://jakarta.ee/specifications/mail/2.1/apidocs/jakarta.mail/jakarta/mail/util/streamprovider[jakarta.mail.util.StreamProvider], that can be easily obtaining via the j`akarta.mail.Session.getStreamProvider()` method. You can then use the StreamProviders to obtain Encoded/Decoded Streams for any of the encoding/decoding methods mail-2.1 supports.

Reviewer: ```suggestion
To facilitate this improvement, Jakarta Mail 2.1 also introduces the link:https://jakarta.ee/specifications/mail/2.1/apidocs/jakarta.mail/jakarta/mail/util/streamprovider[jakarta.mail.util.StreamProvider]  utility interface, which can be easily obtained with the `jakarta.mail.Session.getStreamProvider()` method. You can then use the `StreamProviders` instance to obtain encoded/decoded streams for any of the encoding/decoding methods that `mail-2.1` supports.
```

>  The mail-2.1 feature can be enabled via the server.xml:

Reviewer: ```suggestion
You can enable the `mail-2.1` feature from your server.xml:
```

[source, xml]
----
  <featureManager>
    <feature>mail-2.1</feature>
  </featureManager>
----

> More information about the Jakarta Mail 2.1 specification can be found at the link:https://eclipse-ee4j.github.io/mail/[Jakarta Mail] project website.

Reviewer: ```suggestion
For more information about the Jakarta Mail 2.1 specification, check out the link:https://eclipse-ee4j.github.io/mail/[Jakarta Mail] project website.
```

// https://github.com/OpenLiberty/open-liberty/issues/22415
[#support]
==  Jakarta EE 10: Support Jakarta Servlet 6.0

> The Open Liberty servlet-6.0 feature is an implementation of the Jakarta EE 10 Servlet 6.0 Specification. It includes a number of new features, specification clarifications as well as deprecation features.

Reviewer: ```suggestion
The Open Liberty servlet-6.0 feature is an implementation of the Jakarta EE 10 Servlet 6.0 specification. It includes a number of new features, specification clarifications, and deprecation features.
```

> Prior to servlet-6.0, there were only a few specific methods that could be set for an application's cookie attributes. There was not an easy way for application to set a new attribute (for example, the recent browser security's SameSite attribute). 

Reviewer: ```suggestion
Before `servlet-6.0`, only a few specific methods could be set for an application's cookie attributes. An application could not easily set a new attribute, such as the recent browser security `SameSite` attribute). 
```

> One of the new features in servlet-6.0 is the ability to set any attribute using a Cookie's setAttribute method. In addition, servlet-6.0 also introduces some new request APIs to aid the application's debugging. Application developer can track or refer to a request using the ServletRequest's getRequestId or obtain details of the network connection that is being used by the request by calling ServletRequest's getServletConnection.

Reviewer: ```suggestion
One of the new features in `servlet-6.0` is the ability to set any attribute using a cookie's `setAttribute` method. In addition, `servlet-6.0` also introduces some new request APIs to aid the application's debugging. Application developers can track or refer to a request using the` ServletRequest` `getRequestId` method or obtain details of the network connection that is being used by the request by calling `ServletRequest' `getServletConnection` method.
```

> You can activate the features incorperating the following in your `server.xml`:

Reviewer: ```suggestion
You can activate the feature by adding `servlet-6.0` to your `server.xml`:
```

[source, xml]
----
   <featureManager>
       <feature>servlet-6.0</feature>
   </featureManager>
----

> The Jakarta Servlet 6.0 Specification and Javadocs can be found link:https://jakarta.ee/specifications/servlet/6.0/[here].

Reviewer: ```suggestion
 For more information, refer to the link:https://jakarta.ee/specifications/servlet/6.0/[the Jakarta Servlet 6.0 Specification and Javadocs].
```

// https://github.com/OpenLiberty/open-liberty/issues/22115
[#connectors]
== Connectors-2.1

> This feature allows Open Liberty users that want to use the latest release of the link:https://github.com/eclipse-ee4j/jca-api[Jakarta resource specification] within their application.

Reviewer: ```suggestion
This feature enables you to use the latest release of the link:https://github.com/eclipse-ee4j/jca-api[Jakarta resource specification]in your application.
```

> The main update to connectors to come out of this feature is the addition of Java generics support to the jakarta.resource.cci.ResourceFactory, jakarta.resource.cci.IndexedRecord, and jakarta.resource.cci.MappedRecord classes 

Reviewer: ```suggestion
The main update to connectors in this feature is the addition of Java generics support to the `jakarta.resource.cci.ResourceFactory`, `jakarta.resource.cci.IndexedRecord`, and `jakarta.resource.cci.MappedRecord` classes 
```

> Open Liberty is keeping our connectors implementation up to date with the latest specification.

Reviewer: ```suggestion
```
I'd remove this

> You can activate the features incorperating the following in your `server.xml`:

Reviewer: ```suggestion
You can activate the feature by adding `connectors-2.1` to your `server.xml`:
```

[source, xml]
----
<server>
 <featureManager>
   <feature>connectors-2.1</feature>
 </featureManager>
</server>
----

All other `resourceAdapter` configurations within server.xml can remain unchanged for current applications.

> You can find out more about this specific update below:

Reviewer: ```suggestion
You can find out more about this specific update with the following links:
```

* link:https://github.com/eclipse-ee4j/jca-api[Spec Project]
* link:https://mvnrepository.com/artifact/jakarta.resource/jakarta.resource-api[Spec API]
* link:https://jakarta.ee/specifications/connectors/2.1/apidocs/jakarta.resource/module-summary.html[Spec Javadoc]

// https://github.com/OpenLiberty/open-liberty/issues/22078
[#persistence]
== Persistence 3.1

> Jakarta Persistence API is a richly featured API which provides object-model approach to persisting, fetching, and modifying data stored on a Relational Database System.

Reviewer: ```suggestion
Jakarta Persistence API is a richly featured API that provides an object-model approach to persisting, fetching, and modifying data stored on a relational database system.
```

> JPA has been around since EE 5, but before JPA you had to use EJB's Bean or Container managed entities, or raw JDBC to interact with database storage.

Reviewer: ```suggestion
JPA has been around since Java EE 5.  To interact with database storage before JPA, you had to use EJB bean or container managed entities, or raw JDBC.
```

> 

Reviewer: is there anything new to say about JPA 3.1? The newest thing here is EE 5, which seems odd

[#run]
=== Try it now 

To try out these features, just update your build tools to pull the Open Liberty All Beta Features package instead of the main release. The beta works with Java SE 19, Java SE 18, Java SE 17, Java SE 11, and Java SE 8.

// // // // // // // //
// In the preceding section:
// Check if a new non-LTS Java SE version is supported that needs to be added to the list (17, 11, and 8 are LTS and will remain for a while)
// // // // // // // //

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
  <groupId>io.openliberty.beta</groupId>
  <artifactId>openliberty-runtime</artifactId>
  <version>22.0.0.11-beta</version>
  <type>pom</type>
</dependency>

> ----

Reviewer: does the regular beta include full InstanOn support? I only ask because we have a dedicated container image for it and why would you need that if the regular beta supports InstantOn?

Reviewer 2: That's a good question which I expect the InstantOn standalone blog post to cover.  In short, kind of.... the liberty runtime (what those maven coordinates are for) does include the checkpoint feature.  However, for InstantOn to be functional, certain libraries have to also be available on the OS, which is why we have the InstantOn specific container images (aka, they provide both the Liberty beta runtime, and the necessary CRIU libraries that the OS needs to use the beta functionality).
